### PR TITLE
bucket archive improvements and limits

### PR DESCRIPTION
**Pull request by asutula**
- Configurable limit to the max rep factor for bucket archives, defaults to 3
- Don't expose `address` as part of archive config... just use the user's only address (we don't allow users to create additional addresses)
- Better json representation of `ArchiveConfig` that is consistent with the JS client. Makes json `ArchiveConfig` data interchangeable and consistent between the two clients.

### CODE REVIEW

File: api/bucketsd/pb/bucketsd.proto [hunk 1]
```diff
@@ -193,10 +193,9 @@ message ArchiveConfig {
     repeated string trusted_miners = 4;
     repeated string country_codes = 5;
     ArchiveRenew renew = 6;
-    string addr = 7;
```
Comment (asutula) [inline api/bucketsd/pb/bucketsd.proto:196]: No more address in the api

File: api/bucketsd/service.go [hunk 2]
```diff
@@ -94,17 +94,18 @@ type ctxKey string
...
+	GatewayURL                string
+	GatewayBucketsHost        string
+	IPFSClient                iface.CoreAPI
+	IPNSManager               *ipns.Manager
+	PowergateClient           *pow.Client
+	PowergateAdminToken       string
+	ArchiveTracker            *archive.Tracker
+	Semaphores                *nutil.SemaphorePool
+	MaxBucketSize             int64
+	MaxBucketArchiveRepFactor int
```
Comment (asutula) [inline api/bucketsd/service.go:108]: This gets set via configuration

File: api/bucketsd/service.go [hunk 3]
```diff
@@ -2349,14 +2355,14 @@ func (s *Service) Archive(ctx context.Context, req *pb.ArchiveRequest) (*pb.Arch
...
 	storageConfig := baseArchiveStorageConfig
 	storageConfig.Cold.Filecoin = toFilConfig(archiveConfig)
-
-	if storageConfig.Cold.Filecoin.Address == "" {
-		// We don't have an address to use, which is the case for a BucketArchive.DefaultArchiveConfig
-		// that is the default value, get the default address from the user
-		storageConfig.Cold.Filecoin.Address = defConfRes.DefaultStorageConfig.Cold.Filecoin.Address
-	}
+	// Get the address from the default storage config for this user.
+	storageConfig.Cold.Filecoin.Address = defConfRes.DefaultStorageConfig.Cold.Filecoin.Address
```
Comment (asutula) [inline api/bucketsd/service.go:2365]: We now always get the address from the user's default Powergate `StorageConfig`, they can't modify this via `ArchiveConfig` anymore.

File: api/bucketsd/service.go [hunk 4]
```diff
@@ -2657,3 +2660,16 @@ func toFilConfig(config *mdb.ArchiveConfig) *userPb.FilConfig {
 		TrustedMiners: config.TrustedMiners,
 	}
 }
+
+func (s *Service) validateArchiveConfig(c *mdb.ArchiveConfig) error {
```
Comment (asutula) [inline api/bucketsd/service.go:2664]: Simple validation of an `ArchiveConfig`

File: buckets/local/archive.go [hunk 5]
```diff
@@ -18,41 +18,39 @@ type ArchiveConfig struct {
...
 
 // ArchiveRenew contains renew configuration for a ArchiveConfig.
 type ArchiveRenew struct {
 	// Enabled indicates that deal-renewal is enabled for this Cid.
-	Enabled bool
+	Enabled bool `json:"enabled"`
 	// Threshold indicates how many epochs before expiring should trigger
 	// deal renewal. e.g: 100 epoch before expiring.
-	Threshold int
+	Threshold int `json:"threshold"`
```
Comment (asutula) [inline buckets/local/archive.go:53]: Normal json formatting that is also consistent with the JS client.

File: mongodb/bucketarchives.go [hunk 6]
```diff
@@ -66,8 +66,6 @@ type ArchiveConfig struct {
 	CountryCodes []string `bson:"country_codes"`
 	// Renew indicates deal-renewal configuration.
 	Renew ArchiveRenew `bson:"renew"`
-	// Addr is the wallet address used to store the data in filecoin
-	Addr string `bson:"addr"`
```
Comment (asutula) [inline mongodb/bucketarchives.go:70]: Previously serialized address values will just be ignored from here on out.

**Comment by asutula** [comment]
I added the rep factor limit here in Hub, but also maybe we should add that to Powergate?

File: cmd/hubd/main.go [hunk 7]
```diff
@@ -98,6 +98,10 @@ var (
 				Key:      "buckets.max_size",
 				DefValue: int64(4 * gib),
 			},
+			"bucketsArchiveMaxRepFactor": {
+				Key:      "buckets.max_archive_rep_factor",
```
Comment (jsign) [inline cmd/hubd/main.go:102]: Nit: s/`max_archive_rep_factor`/`archive_max_rep_factor`?

File: cmd/hubd/main.go [hunk 8]
```diff
@@ -378,6 +386,7 @@ var rootCmd = &cobra.Command{
 
 		// Buckets
 		bucketsMaxSize := config.Viper.GetInt64("buckets.max_size")
+		bucketsArchiveMaxRepFactor := config.Viper.GetInt("buckets.max_archive_rep_factor")
```
Comment (jsign) [inline cmd/hubd/main.go:389]: If above changes makes sense, also change here.

**Review by jsign — APPROVED** [review]
LGTM.

Considering it has also breaking API changes, this PR should fire some major version bump process?

**Comment by jsign** [comment]
> I added the rep factor limit here in Hub, but also maybe we should add that to Powergate?

Sounds like a good safety net and potentially useful for people running Powergate alone.
I wouldn't rush to implement it, but sounds like a good idea.

File: cmd/hubd/main.go (hunk 7 above)
Comment (asutula) [reply cmd/hubd/main.go]: oh yea,

**Pull request merged**